addon-test-support/-private/mock/proxy.js

```diff
@@ -1,39 +1,113 @@
+    !descriptor || descriptor.configurable
+  );
+}
+
+function assertWritableDescriptor(target, key) {
+  const descriptor = Object.getOwnPropertyDescriptor(target, key);
+  if (descriptor && 'configurable' in descriptor && !descriptor.configurable) {
+    assert(
+      `Cannot directly set '${key}' on '${target}', because it is a non-configurable and non-writable property.\nSee https://github.com/kaliber5/ember-window-mock/issues/99`,
+      typeof descriptor.set === 'function' || descriptor.writable
+    );
+  }
+}
+
+export default function proxyFactory(
+  original,
+  { makeHolder = () => ({}), makeDescriptors = () => ({}) } = {}
+) {
+  const config = {
+    original,
+    holder: makeHolder(),
+    descriptors: makeDescriptors(),
+    reset() {
+      config.holder = makeHolder();
+      config.descriptors = makeDescriptors();
+    }
+  };
+
+  const proxy = new Proxy(original, {
+    get(target, key) {
+      if (key in config.holder) {
+        return config.holder[key];
       }
-      if (name in holder) {
-        return holder[name];
+      if (key in config.descriptors) {
+        if (config.descriptors[key].get) {
+          return config.descriptors[key].get.call(original);
+        }
+        return config.descriptors[key].value;
       }
-      if (typeof target[name] === 'function') {
-        return mockFunction(target[name], target);
+      if (typeof target[key] === 'function') {
+        return mockFunction(target[key], target);
       }
-      if (typeof target[name] === 'object' && target[name] !== null) {
-        let proxy = proxyFactory(target[name]);
-        holder[name] = proxy;
+      if (typeof target[key] === 'object' && target[key] !== null) {
+        let proxy = proxyFactory(target[key]);
+        config.holder[key] = proxy;
         return proxy;
       }
-      return target[name];
+      return target[key];
     },
-    set (target, name, value) {
-      holder[name] = value;
+    set(target, key, value) {
+      assertWritableDescriptor(target, key);
+
+      if (key in config.descriptors) {
+        if (config.descriptors[key].set) {
+          config.descriptors[key].set.call(original, value);
+          return true;
+        }
+        if (!config.descriptors[key].writable) {
+          return false;
+        }
+        config.descriptors[key].value = value;
+        return true;
+      }
+
+      config.holder[key] = value;
       return true;
     },
-    has(target, prop) {
-      return prop in holder || prop in target;
+    has(target, key) {
+      return key in config.holder || key in target;
+    },
+    deleteProperty(target, key) {
+      assertWritableDescriptor(target, key);
+      delete config.holder[key];
+      delete target[key];
+      return true;
     },
-    deleteProperty(target, prop) {
-      delete holder[prop];
-      delete target[prop];
+    defineProperty(target, key, descriptor) {
+      assertConfigurableDescriptor(target, key);
```

Comment:
>> Because users would not be able to use Object.defineProperty for document in the real world, we must prohibit it here as well.
>
> Not sure why?

Turns out, you can still modify a non-configurable descriptor. 😂 
You can flip the `writable` flag for instance. But you can't switch from `value` to `get` / `set`. It's probably safest, if we copy the descriptor from `target` over to `holder` and then try mutating it. If this throws an error, we know it's not allowed.

```js
const obj = {};

Object.defineProperty(obj, 'foo', { configurable: false, writable: true, value: 'bar' });

// works
Object.defineProperty(obj, 'foo', { configurable: false, writable: false, value: 'quux' });

// breaks
Object.defineProperty(obj, 'foo', { configurable: false, get: () => 'baz', set: undefined });
```

Reply:
Ah, but this would then "poison" the descriptor on the `holder` as well.

Reply:
It's now checked using a dummy object.

Reply:
First, sorry for letting this wait for so long! Was a tough week... 

I am still not really understanding why we cannot just define the property on the holder? In what case would that throw an error? 

Your examples above are referring to defining a property an a regular object, that already has a descriptor. But do those examples apply also when we define the property on a holder of a Proxy, instead of the target itself?